### PR TITLE
VersionCatalog: Avoid accessor name conflicting on Windows

### LINKED ISSUE #20060: Version catalogs run into unexpected namespace conflicts
I've noticed that when parsing a version catalog, the namespace `x.y` will result in some type of conflict with the namespace `xy`.
Concretely, when I have the following two lines in my version catalog toml file:
```
com-googlecode-json-simple = "com.googlecode.json-simple:json-simple:1.1.1"
com-google-code-gson = "com.google.code.gson:gson:2.8.2"
```
The second line fails with the error:

```
> Could not create an instance of type org.gradle.accessors.dm.LibrariesForLibs.
   > org/gradle/accessors/dm/LibrariesForLibs$ComGooglecodeLibraryAccessors (wrong name: org/gradle/accessors/dm/LibrariesForLibs$ComGoogleCodeLibraryAccessors)
```

The lowercase "code" in `ComGooglecodeLibraryAccessors` suggests that it's searching in the `googlecode` namespace and not `google-code`.

I've done my best to verify that this behavior shows up if and only if I have both `xy` and `x.y` in the toml file.

### Expected Behavior
Gradle can parse this version catalog without error.

### Current Behavior
Gradle runs into the error described above.

### Context
We noticed this error while migrating to using version catalogs.
Firstly, the error message was vague and hard to debug.
Secondly, this requires us to manually edit these namespace "conflicts" which shouldn't be really conflicts.

### Steps to Reproduce 
This error is reproducible with my following sample repo: https://github.com/lchen8/gradle-bug-test-application/
It fails upon trying to sync/config with Gradle.
Within the sample repo, I've commented in [libs.versions.toml L9](https://github.com/lchen8/gradle-bug-test-application/blob/main/gradle/libs.versions.toml#L9) where the problem stems from.

### Your Environment
This problem appears both during Gradle sync (from Android studio) and any gradle command from terminal
Edited to add: We are using Gradle 7.4

Build scan URL: (The build scan was not published due to a configuration problem.)

### PR DESCRIPTION
**Pull request by scaventz**
Take this TOML file for example
```
[libraries]
com-company-libs-a = "com.company:libs-a:1.0"
com-companylibs-b = "com.companylibs:libs-b:1.0"
```

Gradle generates these two library accessor:
```
ComCompanyLibsLibraryAccessors
ComCompanylibsLibraryAccessors
```
On a case-insensitive platform, like Windows, the first one will be overwritten by the second one.
<strike>
To avoid this, we can adjust the name generation rules to introduce a `$` symbol between parent and child accessors into class name:

```
Com$Company$LibsLibraryAccessors
Com$CompanylibsLibraryAccessors
```
</strike>

To avoid this, we can adjust the name generation rules to introduce a `$` into class name:
```
ComCompanylibsLibraryAccessors
ComCompanyLibsLibraryAccessors$1
```

Signed-off-by: Xin Wang <scaventz@outlook.com>

Fixes #20060

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by donat** [comment]
I've looked at your solution. What worries me is that you add `$` symbols for _all_ generated classes for _all_ dependency version catalog users. Is there a way to add `$`s when resolving naming conflicts?

**Comment by scaventz** [comment]
Updated, with both commit and description of this PR :)

**Review by jbartok — COMMENTED** [review]
What happens when the conflicting aliases are declared in different catalogs? Would be good to have a test for that.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/catalog/LibrariesSourceGenerator.java [hunk 1]
```diff
@@ -204,10 +207,30 @@ private void writeBundleAccessorClass(ClassNode classNode) throws IOException {
         writeLn();
     }
 
+    private final Map<String, Integer> classNameCounter = new HashMap<>();
```
Comment (jbartok) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/catalog/LibrariesSourceGenerator.java:210]: Field declarations should be before the methods.

**Comment by scaventz** [comment]
> What happens when the conflicting aliases are declared in different catalogs? Would be good to have a test for that.

Sorry for late response, do you mean something like: 
```Groovy
dependencyResolutionManagement {
  versionCatalogs {
      create("libs") {
          library("com-company-libs-a", "com.company:libs-a:1.0")
      }

      create("moreLibs") {
          library("com-companylibs-b", "com.companylibs:libs-b:1.0")
      }
  }
}
```
It doesn't seem relevant to this issue to me, because there will be two classes generated for above case:

LibrariesForLibs
```
public class LibrariesForLibs extends AbstractExternalDependencyFactory {
    public static class ComLibraryAccessors ...
    public static class ComCompanyLibraryAccessors ...
    public static class ComCompanyLibsLibraryAccessors ...
}
```

LibrariesForMoreLibs
```
public class LibrariesForMoreLibs extends AbstractExternalDependencyFactory {
    public static class ComLibraryAccessors ...
    public static class ComCompanylibsLibraryAccessors ...
}
```

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/catalog/LibrariesSourceGenerator.java (hunk 1 above)
Comment (scaventz) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/catalog/LibrariesSourceGenerator.java]: Thanks for the review, updated

**Comment by jbartok** [comment]
It would be good to have a test to confirm and lock in place this behavior.

**Comment by scaventz** [comment]
> It would be good to have a test to confirm and lock in place this behavior.

Sorry I forgot to mention that there is separate commit(a3d9dd12c6) of this PR containing a new test to cover this behavior. I wonder if it addresses our concern?

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test and merge

**Pull request merged**